src/unix/fs.c

```diff
@@ -698,10 +697,89 @@ static ssize_t uv__fs_sendfile(uv_fs_t* req) {
 
 
 static ssize_t uv__fs_utime(uv_fs_t* req) {
-  struct utimbuf buf;
-  buf.actime = req->atime;
-  buf.modtime = req->mtime;
-  return utime(req->path, &buf); /* TODO use utimes() where available */
+#if defined(__linux__)
+  /* utimesat() has nanosecond resolution but we stick to microseconds
+   * for the sake of consistency with other platforms.
+   */
+  static int no_utimesat;
+  struct timespec ts[2];
+  struct timeval tv[2];
+  int r;
+
+  if (no_utimesat)
+    goto skip;
+
+  ts[0].tv_sec  = req->atime;
+  ts[0].tv_nsec = (uint64_t)(req->atime * 1000000) % 1000000 * 1000;
+  ts[1].tv_sec  = req->mtime;
+  ts[1].tv_nsec = (uint64_t)(req->mtime * 1000000) % 1000000 * 1000;
+
+  r = uv__utimesat(AT_FDCWD, req->path, ts, 0);
+  if (r == 0)
+    return r;
+
+  if (errno != ENOSYS)
+    return r;
+
+  no_utimesat = 1;
+
+skip:
+
+  tv[0].tv_sec  = req->atime;
+  tv[0].tv_usec = (uint64_t)(req->atime * 1000000) % 1000000;
+  tv[1].tv_sec  = req->mtime;
+  tv[1].tv_usec = (uint64_t)(req->mtime * 1000000) % 1000000;
+
+  r = utimes(req->path, tv);
+  if (r == 0)
+    return r;
+
+  switch (errno) {
+  case ENOENT:
+    /* if (fcntl(req->file, F_GETFL) == -1 && errno == EBADF) */
+      break;
+    /* Fall through. */
+
+  case EACCES:
+  case ENOTDIR:
+    errno = ENOSYS;
+    break;
+  }
+
+  return r;
+
+#elif defined(__APPLE__)                                                      \
+    || defined(__DragonFly__)                                                 \
+    || defined(__FreeBSD__)                                                   \
+    || defined(__FreeBSD_kernel__)                                            \
+    || defined(__NetBSD__)                                                    \
+    || defined(__OpenBSD__)
+  struct timeval tv[2];
+  tv[0].tv_sec  = req->atime;
+  tv[0].tv_usec = (uint64_t)(req->atime * 1000000) % 1000000;
+  tv[1].tv_sec  = req->mtime;
+  tv[1].tv_usec = (uint64_t)(req->mtime * 1000000) % 1000000;
+  return utimes(req->path, tv);
+#elif defined(_AIX71)                                                         \
```

Comment:
> What's AIX error code -109? `Function not implemented`?

Ah could very well be ENOSYS. The CI AIX machines are AIX 6.1 so I'd guess `_AIX71` might not be defined. Maybe the test also needs the #ifdef (for comparison, see https://github.com/libuv/libuv/pull/811 where this was done for `uv__fs_futime`).

Reply:
@richardlau Think I should I fall back to using `utime()` on `_AIX && ! _AIX71` then?

Reply:
@Fishrock123 libuv v1.x's minimum supported AIX is 6 (https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md) so yes, I think that sounds sensible. 

cc @libuv/aix

Reply:
@richardlau Doesn't look like that team mention worked. :)

Reply:
I think you need to be a member of the org (libuv) for mentions to work 😞